main/src/library/RedBlackTree.flix

```diff
@@ -1,23 +1,33 @@
-///
-/// The color of a red-black tree node.
-///
-enum Color {
-    case Red
-    case Black
-    case DoubleBlack
-}
-
-///
-/// The red-black tree type.
-///
-enum RedBlackTree[t] {
-    case Leaf,
-    case DoubleLeaf,
-    case Node(Color, RedBlackTree[t], t, RedBlackTree[t])
-}
-
 namespace RedBlackTree {
 
+    ///
+    /// The red-black tree used by `Set` and `Map`.
+    ///
+    /// A red-black tree is a balanced binary tree.
+    /// Each node is either red or black, although a
+    /// transitory color double-black is allowed during deletion.
+    /// Furthermore, the red-black tree satisfy the following
+    /// two balance invariants.
+    ///
+    /// 1. No red node has a red parent.
+    /// 2. Every path from the root to a leaf contains the same
+    /// number of black nodes.
```

Comment:
Question: Is this enought (answer: no). Can you discover what is missing? :) (Hint: It has nothing to do with colors)

Reply:
Are you thinking about the number of children?

Reply:
No :) What is it that makes e.g. insert work correctly in addition to the two properties above?

Reply:
Ordering?

Reply:
The search tree invariant!

Reply:
I guess my idea was that it was implicit because I mention that is a binary tree, would you like it to list that the elements must have some total order?

Reply:
A "binary tree" does not have to satisfy the "search tree property". It is custom to both write "binary search tree" and to state the property explicitly, i.e. if node n has key v then such and such for the left and right trees.

Reply:
Ahh yes, that had slipped my mind.